Add builtin time, geometric and most range types

wolframm opened the pull request:
I added some builtin types as discussed in issue #293.

This is the first time I contribute to a project on Github, so I'm more than happy to fix any issues you might see with pull request.

Once is merged, I would add `numrange`, but I wanted to ask if it wouldn't be better to change the current `numeric` type from `GenericType<Object>(TypeOid.numeric)` that accepts `int`, `double` and `String` to `GenericType<Decimal>(TypeOid.numeric)` using the decimal library (https://pub.dev/packages/decimal).

isoos commented:
Thanks for the PR! I'll have time for review in the coming days, but a quick look:
- code analysis CI test is failing, please check the lints/warnings/errors and fix them
- please bump the version and add a few bits about the PR in `CHANGELOG.md`
- It is easier to review/iterate on smaller PRs, e.g. one or just a few related type / PR. If it is not that hard to split, I would be very happy to review more smaller PRs than a large one. (If not no worries, it just make take longer).

wolframm commented:
I ran `dart test` and fixed one failing test case for `daterange`. Sorry about that.

I bumped the version and added a note to changelog.

I ran `dart analyze` and `dart test` and no issues were found at least on my setup, so hopefully everything should work now.

The most tricky part of the code were the range types. I think I covered most edge cases in `encoding_test.dart`, but this would be the best place to try to "brake" my code. Geotypes and time seemed rather simple.

wolframm commented:
I just upgraded my Dart version to `3.3.0` and now I also see one of the warnings that your CI flow reported:
```
warning • lib/src/messages/server_messages.dart:412:17 • The parameter type of '==' operators should be non-nullable. Try using a non-nullable type. • non_nullable_equals_parameter
```
I made no changes to `server_messages.dart`. Do you want to make the necessary fix there?

isoos commented:
> Could you make the necessary fix in server_messages.dart (or ignore the warning for now)?

Good catch! Pushed a fix to the git repo, you may rebase to it (or just adapt the same change).

isoos commented on CHANGELOG.md:
```diff
@@ -1,13 +1,22 @@
 ## 3.0.7
 
-- Allow cleartext passwords when secure connection is used. ([#283](https://github.com/isoos/postgresql-dart/pull/283) by [simolus3](https://github.com/simolus3))
+- Allow cleartext passwords when secure connection is
```
Are these some kind of auto-formatting of the markdown file? I wouldn't mind that outside of this PR doing only just that formatting, but I'd prefer that this PR reverts the rest of the changes here.

wolframm replied:
I added a comment explaining the constructor and made switch statement shorter. I'm unfamiliar with bit operations so I would prefer to leave it as is for now.

isoos commented on lib/postgres.dart:
```diff
@@ -17,6 +17,8 @@ export 'src/exceptions.dart';
 export 'src/replication.dart';
 export 'src/types.dart';
+export 'src/types/geo_types.dart';
+export 'src/types/range_types.dart' hide Range, DiscreteRange, ContinuousRange;
```
Could these types be private types, e.g. `_Range`, `_DiscreteRange` and `_ContinuousRange`, so that it is explicit in the code that they are not intended to be public.

On the other hand `Range` has useful public methods, maybe it should be also public? I'm not sure why would we hide these?

wolframm replied:
removed `hide` expression

isoos commented on lib/src/types.dart:
```diff
@@ -259,6 +339,27 @@ abstract class Type<T extends Object> {
+  static const int8range = GenericType<IntRange>(TypeOid.int8range);
+
+  /// Must be a [DateRange]
+  static const daterange = GenericType<DateRange>(TypeOid.daterange);
```
rename `daterange` to `dateRange` (also in `TypeOid`)

isoos commented on lib/src/types.dart:
```diff
@@ -227,6 +289,24 @@ abstract class Type<T extends Object> {
+  static const line = GenericType<Line>(TypeOid.line);
+
+  /// Must be a [Lseg]
+  static const lseg = GenericType<Lseg>(TypeOid.lseg);
```
I'm a bit undecided if this should be `lseg` or `lineSegment`. I think it may be worth to expand the name (also in `TypeOid`).

wolframm replied:
Changed everything to lineSegment/LineSegment

isoos commented on lib/src/types/binary_codec.dart:
```diff
@@ -505,6 +628,35 @@ class PostgresBinaryEncoder {
+      case (false, true) || (true, false):
+        final value = range.lower ?? range.upper!;
+        _encodeRangeValue(value, encoding, elementTypeOid, buffer);
+      case (true, true):
```
let's add explicit `break` here and before the `case` too.

wolframm replied:
done

isoos commented on lib/src/types/range_types.dart:
```diff
@@ -0,0 +1,258 @@
+
+  Bounds.fromFlag(int flag) {
+    switch (flag) {
+      case 0 || 1:
```
This switch-case structure is a bit obscure, what is the source of it? Can we do bit operations instead?

wolframm replied:
I added a comment explaining the constructor and made switch statement shorter. I'm unfamiliar with bit operations so I would prefer to leave it as is for now.

isoos commented on lib/src/types/type_registry.dart:
```diff
@@ -19,27 +18,40 @@ class TypeOid {
   static const character = 1042;
+  static const circle = 718;
   static const date = 1082;
+  static const daterange = 3912;
```
I also think this should be `dateRange` (also in `types.dart`)

isoos commented on lib/src/types/type_registry.dart:
```diff
@@ -19,27 +18,40 @@ class TypeOid {
+  static const time = 1083;
   static const timestampWithoutTimezone = 1114;
   static const timestampWithTimezone = 1184;
+  static const tsrange = 3908;
```
let's rename this to `timestampRange` for consistency with the above types (also in `types.dart`). Hm, maybe `timestampWithoutTimezoneRange`, but better to use the shorter name.

isoos commented on lib/src/types/type_registry.dart:
```diff
@@ -19,27 +18,40 @@ class TypeOid {
   static const timestampWithoutTimezone = 1114;
   static const timestampWithTimezone = 1184;
+  static const tsrange = 3908;
+  static const tstzrange = 3910;
```
This could be `timestampTzRange` instead of `timestampWithTimezoneRange` (also in `types.dart`). We should alias/migrate the `timestampWithTimezone` to `timestampTz` independently to this PR.

isoos reviewed (changes requested):
This looks awesome! I have mostly nits and a few rename requests, plus a few clarification questions.

wolframm commented:
Is there anything you want me to do regarding the Codecov report? 

All the issues flagged regarding the `==`-operator should be covered by `expectReversible` function in `encoding_test.dart`. Or perhaps I misunderstand what 'expectReversible` is doing.

isoos commented:
Thanks for the updates! I'll merge this now, and play with it locally, maybe refactor a bit here and there. I am planning to publish the new version with this later today or maybe tomorrow at latest.
